pkg/controller/highperf/highperf_controller.go

```diff
@@ -0,0 +1,234 @@
+	"k8s.io/klog/v2"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+
+	networkclientset "k8s.io/cloud-provider-gcp/crd/client/network/clientset/versioned"
+	nodeutil "k8s.io/cloud-provider-gcp/pkg/util"
+	"k8s.io/cloud-provider-gcp/providers/gce"
+	controllersmetrics "k8s.io/component-base/metrics/prometheus/controllers"
+
+	networkv1 "k8s.io/cloud-provider-gcp/crd/apis/network/v1"
+	networkv1alpha1 "k8s.io/cloud-provider-gcp/crd/apis/network/v1alpha1"
+)
+
+type Controller struct {
+	gceCloud         *gce.Cloud
+	networkClientset networkclientset.Interface
+
+	nwInformer   cache.SharedIndexInformer
+	gnpInformer  cache.SharedIndexInformer
+	nodeInformer coreinformers.NodeInformer
+
+	queue workqueue.RateLimitingInterface
+}
+
+func NewHighPerfController(
+	gceCloud *gce.Cloud,
+	networkClientset networkclientset.Interface,
+	nwInformer cache.SharedIndexInformer,
+	gnpInformer cache.SharedIndexInformer,
+	nodeInformer coreinformers.NodeInformer) *Controller {
+
+	hc := &Controller{
+		gceCloud:         gceCloud,
+		networkClientset: networkClientset,
+		nwInformer:       nwInformer,
+		gnpInformer:      gnpInformer,
+		nodeInformer:     nodeInformer,
+		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "highperf"),
+	}
+
+	// TODO (bmorettodama): Add event handlers for Network and GNP object DeleteFunc
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    nodeutil.CreateAddNodeHandler(hc.nodeAdd),
+		UpdateFunc: nodeutil.CreateUpdateNodeHandler(hc.nodeUpdate),
+	})
+
+	return hc
+}
+
+func (c *Controller) Run(numWorkers int, stopCh <-chan struct{}, controllerManagerMetrics *controllersmetrics.ControllerManagerMetrics) {
+	defer utilruntime.HandleCrash()
+
+	ctx, cancelFn := context.WithCancel(context.Background())
+	defer cancelFn()
+	defer c.queue.ShutDown()
+
+	klog.Info("Starting high perf controller")
+	defer klog.Info("Shutting down high perf controller")
+	controllerManagerMetrics.ControllerStarted("highperf")
+	defer controllerManagerMetrics.ControllerStopped("highperf")
+
+	if !cache.WaitForNamedCacheSync("node", stopCh, c.nodeInformer.Informer().HasSynced) {
+		return
+	}
+
+	if !cache.WaitForNamedCacheSync("networks", stopCh, c.nwInformer.HasSynced) {
+		return
+	}
+
+	if !cache.WaitForNamedCacheSync("gkenetworkparamset", stopCh, c.gnpInformer.HasSynced) {
+		return
+	}
+
+	for i := 0; i < numWorkers; i++ {
+		go wait.UntilWithContext(ctx, c.runWorker, time.Second)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) runWorker(ctx context.Context) {
+	for c.processNextItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextItem(ctx context.Context) bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.queue.Done(key)
+
+	c.processNode(ctx, key.(string))
+	// TODO (bmorettodama): handle errors from processNode (try again)
```

Comment:
not TODO, just add it ;) you have an example here 

https://github.com/kubernetes/sample-controller/blob/master/controller.go

and here https://github.com/kubernetes/client-go/blob/master/examples/workqueue/main.go#L53-L69